File: csharp-version/VidyoPasswordManager.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Net.Http;
using System.ServiceModel;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml;

namespace VidyoPasswordManager
{
    public partial class MainForm : Form
    {
        private readonly string[] monthUsernames = {
            "jan", "feb", "mar", "apr", "may", "jun",
            "jul", "aug", "sep", "oct", "nov", "dec"
        };

        private ComboBox cmbUsername;
        private TextBox txtPassword;
        private TextBox txtPortalUrl;
        private TextBox txtAdminUser;
        private TextBox txtAdminPassword;
        private Button btnUpdatePassword;
        private Button btnBulkUpdate;
        private CheckedListBox clbUsers;
        private TextBox txtLog;
        private ProgressBar progressBar;

        public MainForm()
        {
            InitializeComponent();
            LoadMonthUsernames();
        }

        private void InitializeComponent()
        {
            this.Text = "Vidyo Password Manager";
            this.Size = new Size(600, 500);
            this.StartPosition = FormStartPosition.CenterScreen;

            // Portal Configuration Group
            var grpPortal = new GroupBox
            {
                Text = "Portal Configuration",
                Location = new Point(10, 10),
                Size = new Size(560, 120)
            };

            var lblPortalUrl = new Label { Text = "Portal URL:", Location = new Point(10, 25), Size = new Size(80, 20) };
            txtPortalUrl = new TextBox { Location = new Point(100, 22), Size = new Size(300, 20), Text = "http://your-vidyo-portal.com" };

            var lblAdminUser = new Label { Text = "Admin User:", Location = new Point(10, 50), Size = new Size(80, 20) };
            txtAdminUser = new TextBox { Location = new Point(100, 47), Size = new Size(150, 20) };

            var lblAdminPassword = new Label { Text = "Admin Pass:", Location = new Point(10, 75), Size = new Size(80, 20) };
            txtAdminPassword = new TextBox { Location = new Point(100, 72), Size = new Size(150, 20), UseSystemPasswordChar = true };

            grpPortal.Controls.AddRange(new Control[] { lblPortalUrl, txtPortalUrl, lblAdminUser, txtAdminUser, lblAdminPassword, txtAdminPassword });

            // Single User Update Group
            var grpSingle = new GroupBox
            {
                Text = "Single User Password Update",
                Location = new Point(10, 140),
                Size = new Size(280, 120)
            };

            var lblUsername = new Label { Text = "Username:", Location = new Point(10, 25), Size = new Size(70, 20) };
            cmbUsername = new ComboBox { Location = new Point(85, 22), Size = new Size(80, 20), DropDownStyle = ComboBoxStyle.DropDownList };

            var lblPassword = new Label { Text = "New Password:", Location = new Point(10, 50), Size = new Size(90, 20) };
            txtPassword = new TextBox { Location = new Point(105, 47), Size = new Size(150, 20), UseSystemPasswordChar = true };

            btnUpdatePassword = new Button { Text = "Update Password", Location = new Point(105, 80), Size = new Size(120, 30) };
            btnUpdatePassword.Click += BtnUpdatePassword_Click;

            grpSingle.Controls.AddRange(new Control[] { lblUsername, cmbUsername, lblPassword, txtPassword, btnUpdatePassword });

            // Bulk Update Group
            var grpBulk = new GroupBox
            {
                Text = "Bulk Password Update",
                Location = new Point(300, 140),
                Size = new Size(270, 120)
            };

            var lblBulkPassword = new Label { Text = "Password for all:", Location = new Point(10, 25), Size = new Size(90, 20) };
            var txtBulkPassword = new TextBox { Location = new Point(105, 22), Size = new Size(150, 20), UseSystemPasswordChar = true, Name = "txtBulkPassword" };

            clbUsers = new CheckedListBox { Location = new Point(10, 50), Size = new Size(150, 60) };
            
            btnBulkUpdate = new Button { Text = "Bulk Update", Location = new Point(170, 80), Size = new Size(90, 30) };
            btnBulkUpdate.Click += BtnBulkUpdate_Click;

            grpBulk.Controls.AddRange(new Control[] { lblBulkPassword, txtBulkPassword, clbUsers, btnBulkUpdate });

            // Progress and Log
            progressBar = new ProgressBar { Location = new Point(10, 270), Size = new Size(560, 20), Visible = false };

            var lblLog = new Label { Text = "Operation Log:", Location = new Point(10, 300), Size = new Size(100, 20) };
            txtLog = new TextBox 
            { 
                Location = new Point(10, 325), 
                Size = new Size(560, 120), 
                Multiline = true, 
                ScrollBars = ScrollBars.Vertical,
                ReadOnly = true
            };

            this.Controls.AddRange(new Control[] { grpPortal, grpSingle, grpBulk, progressBar, lblLog, txtLog });
        }

        private void LoadMonthUsernames()
        {
            cmbUsername.Items.AddRange(monthUsernames);
            clbUsers.Items.AddRange(monthUsernames);
        }

        private async void BtnUpdatePassword_Click(object sender, EventArgs e)
        {
            if (string.IsNullOrEmpty(cmbUsername.Text) || string.IsNullOrEmpty(txtPassword.Text))
            {
                MessageBox.Show("Please select a username and enter a password.", "Validation Error");
                return;
            }

            await UpdateSinglePassword(cmbUsername.Text, txtPassword.Text);
        }

        private async void BtnBulkUpdate_Click(object sender, EventArgs e)
        {
            var selectedUsers = clbUsers.CheckedItems.Cast<string>().ToList();
            var bulkPasswordTextBox = this.Controls.Find("txtBulkPassword", true).FirstOrDefault() as TextBox;
            
            if (selectedUsers.Count == 0 || string.IsNullOrEmpty(bulkPasswordTextBox?.Text))
            {
                MessageBox.Show("Please select users and enter a password.", "Validation Error");
                return;
            }

            await UpdateMultiplePasswords(selectedUsers, bulkPasswordTextBox.Text);
        }

        private async Task UpdateSinglePassword(string username, string password)
        {
            try
            {
                LogMessage($"Starting password update for user: {username}");
                
                // TODO: Implement actual Vidyo API call
                var success = await CallVidyoUpdateMemberAPI(username, password);
                
                if (success)
                {
                    LogMessage($"✓ Password updated successfully for {username}");
                    txtPassword.Clear();
                }
                else
                {
                    LogMessage($"✗ Failed to update password for {username}");
                }
            }
            catch (Exception ex)
            {
                LogMessage($"✗ Error updating {username}: {ex.Message}");
            }
        }

        private async Task UpdateMultiplePasswords(List<string> usernames, string password)
        {
            progressBar.Visible = true;
            progressBar.Maximum = usernames.Count;
            progressBar.Value = 0;

            foreach (var username in usernames)
            {
                await UpdateSinglePassword(username, password);
                progressBar.Value++;
                Application.DoEvents(); // Keep UI responsive
            }

            progressBar.Visible = false;
            LogMessage($"Bulk update completed for {usernames.Count} users.");
        }

        private async Task<bool> CallVidyoUpdateMemberAPI(string username, string password)
        {
            // TODO: Implement actual SOAP call to Vidyo API
            // This is a placeholder - you'll need to add service reference to Vidyo WSDL
            
            await Task.Delay(1000); // Simulate API call
            
            // Example of what the actual implementation would look like:
            /*
            try
            {
                var client = new VidyoPortalAdminServiceClient();
                client.ClientCredentials.UserName.UserName = txtAdminUser.Text;
                client.ClientCredentials.UserName.Password = txtAdminPassword.Text;
                
                // Get member first
                var member = await client.GetMemberAsync(username);
                
                // Update password
                member.password = password;
                var result = await client.UpdateMemberAsync(member.memberID, member);
                
                return result.Equals("OK");
            }
            catch (Exception ex)
            {
                LogMessage($"API Error: {ex.Message}");
                return false;
            }
            */
            
            return true; // Placeholder return
        }

        private void LogMessage(string message)
        {
            var timestamp = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
            txtLog.AppendText($"[{timestamp}] {message}\r\n");
            txtLog.ScrollToCaret();
        }
    }

    // Program entry point
    public static class Program
    {
        [STAThread]
        public static void Main()
        {
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new MainForm());
        }
    }
}

File: python-version/vidyo_password_manager.py
import tkinter as tk
from tkinter import ttk, messagebox, scrolledtext
import requests
from requests.auth import HTTPBasicAuth
import xml.etree.ElementTree as ET
import asyncio
import threading
from datetime import datetime
import base64

class VidyoPasswordManager:
    def __init__(self, root):
        self.root = root
        self.root.title("Vidyo Password Manager")
        self.root.geometry("700x600")
        
        self.month_usernames = [
            "jan", "feb", "mar", "apr", "may", "jun",
            "jul", "aug", "sep", "oct", "nov", "dec"
        ]
        
        self.setup_ui()
        
    def setup_ui(self):
        # Portal Configuration Frame
        portal_frame = ttk.LabelFrame(self.root, text="Portal Configuration", padding="10")
        portal_frame.grid(row=0, column=0, columnspan=2, sticky="ew", padx=10, pady=5)
        
        ttk.Label(portal_frame, text="Portal URL:").grid(row=0, column=0, sticky="w", padx=5)
        self.portal_url = ttk.Entry(portal_frame, width=40)
        self.portal_url.grid(row=0, column=1, padx=5)
        self.portal_url.insert(0, "http://your-vidyo-portal.com")
        
        ttk.Label(portal_frame, text="Admin User:").grid(row=1, column=0, sticky="w", padx=5)
        self.admin_user = ttk.Entry(portal_frame, width=20)
        self.admin_user.grid(row=1, column=1, sticky="w", padx=5)
        
        ttk.Label(portal_frame, text="Admin Password:").grid(row=2, column=0, sticky="w", padx=5)
        self.admin_password = ttk.Entry(portal_frame, width=20, show="*")
        self.admin_password.grid(row=2, column=1, sticky="w", padx=5)
        
        # Single User Update Frame
        single_frame = ttk.LabelFrame(self.root, text="Single User Password Update", padding="10")
        single_frame.grid(row=1, column=0, sticky="ew", padx=10, pady=5)
        
        ttk.Label(single_frame, text="Username:").grid(row=0, column=0, sticky="w", padx=5)
        self.username_combo = ttk.Combobox(single_frame, values=self.month_usernames, state="readonly", width=10)
        self.username_combo.grid(row=0, column=1, padx=5)
        
        ttk.Label(single_frame, text="New Password:").grid(row=1, column=0, sticky="w", padx=5)
        self.password_entry = ttk.Entry(single_frame, width=20, show="*")
        self.password_entry.grid(row=1, column=1, padx=5)
        
        self.update_btn = ttk.Button(single_frame, text="Update Password", command=self.update_single_password)
        self.update_btn.grid(row=2, column=1, pady=10)
        
        # Bulk Update Frame
        bulk_frame = ttk.LabelFrame(self.root, text="Bulk Password Update", padding="10")
        bulk_frame.grid(row=1, column=1, sticky="ew", padx=10, pady=5)
        
        ttk.Label(bulk_frame, text="Password for all:").grid(row=0, column=0, sticky="w", padx=5)
        self.bulk_password = ttk.Entry(bulk_frame, width=20, show="*")
        self.bulk_password.grid(row=0, column=1, padx=5)
        
        ttk.Label(bulk_frame, text="Select Users:").grid(row=1, column=0, sticky="nw", padx=5)
        
        # Create frame for checkboxes
        checkbox_frame = ttk.Frame(bulk_frame)
        checkbox_frame.grid(row=1, column=1, padx=5, pady=5)
        
        self.user_vars = {}\n        for i, username in enumerate(self.month_usernames):
            var = tk.BooleanVar()
            self.user_vars[username] = var
            cb = ttk.Checkbutton(checkbox_frame, text=username, variable=var)
            cb.grid(row=i//3, column=i%3, sticky="w", padx=2)
        
        self.bulk_update_btn = ttk.Button(bulk_frame, text="Bulk Update", command=self.update_bulk_passwords)
        self.bulk_update_btn.grid(row=2, column=1, pady=10)
        
        # Progress Bar
        self.progress = ttk.Progressbar(self.root, mode='determinate')
        self.progress.grid(row=2, column=0, columnspan=2, sticky="ew", padx=10, pady=5)
        
        # Log Frame
        log_frame = ttk.LabelFrame(self.root, text="Operation Log", padding="10")
        log_frame.grid(row=3, column=0, columnspan=2, sticky="ew", padx=10, pady=5)
        
        self.log_text = scrolledtext.ScrolledText(log_frame, height=15, width=80)
        self.log_text.grid(row=0, column=0, sticky="ew")
        
        # Configure grid weights
        self.root.columnconfigure(0, weight=1)
        self.root.columnconfigure(1, weight=1)
        log_frame.columnconfigure(0, weight=1)
        
    def log_message(self, message):
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        self.log_text.insert(tk.END, f"[{timestamp}] {message}\n")
        self.log_text.see(tk.END)
        self.root.update_idletasks()
        
    def validate_inputs(self):
        if not self.portal_url.get() or not self.admin_user.get() or not self.admin_password.get():
            messagebox.showerror("Validation Error", "Please fill in all portal configuration fields.")
            return False
        return True
        
    def update_single_password(self):
        if not self.validate_inputs():
            return
            
        username = self.username_combo.get()
        password = self.password_entry.get()
        
        if not username or not password:
            messagebox.showerror("Validation Error", "Please select a username and enter a password.")
            return
            
        # Run in thread to prevent UI blocking
        threading.Thread(target=self._update_password_thread, args=(username, password), daemon=True).start()
        
    def update_bulk_passwords(self):
        if not self.validate_inputs():
            return
            
        selected_users = [user for user, var in self.user_vars.items() if var.get()]
        password = self.bulk_password.get()
        
        if not selected_users or not password:
            messagebox.showerror("Validation Error", "Please select users and enter a password.")
            return
            
        # Run in thread to prevent UI blocking
        threading.Thread(target=self._bulk_update_thread, args=(selected_users, password), daemon=True).start()
        
    def _update_password_thread(self, username, password):
        try:
            self.log_message(f"Starting password update for user: {username}")
            success = self.call_vidyo_api(username, password)
            
            if success:
                self.log_message(f"✓ Password updated successfully for {username}")
                self.password_entry.delete(0, tk.END)
            else:
                self.log_message(f"✗ Failed to update password for {username}")
        except Exception as e:
            self.log_message(f"✗ Error updating {username}: {str(e)}")
            
    def _bulk_update_thread(self, usernames, password):
        self.progress['maximum'] = len(usernames)
        self.progress['value'] = 0
        
        for i, username in enumerate(usernames):
            self._update_password_thread(username, password)
            self.progress['value'] = i + 1
            self.root.update_idletasks()
            
        self.log_message(f"Bulk update completed for {len(usernames)} users.")
        
    def call_vidyo_api(self, username, password):
        """
        Call Vidyo UpdateMember API to change user password
        This is a placeholder implementation - you'll need to implement the actual SOAP call
        """
        try:
            # TODO: Implement actual SOAP call to Vidyo API
            # For now, this is a simulation
            
            portal_url = self.portal_url.get()
            admin_user = self.admin_user.get()
            admin_pass = self.admin_password.get()
            
            # Example of what the actual implementation would look like:
            """
            # Construct SOAP envelope for UpdateMember
            soap_envelope = f'''<?xml version="1.0" encoding="utf-8"?>
            <soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
                <soap:Header>
                    <wsse:Security xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
                        <wsse:UsernameToken>
                            <wsse:Username>{admin_user}</wsse:Username>
                            <wsse:Password>{admin_pass}</wsse:Password>
                        </wsse:UsernameToken>
                    </wsse:Security>
                </soap:Header>
                <soap:Body>
                    <UpdateMember xmlns="http://portal.vidyo.com/admin">
                        <memberID>{username}</memberID>
                        <member>
                            <name>{username}</name>
                            <password>{password}</password>
                        </member>
                    </UpdateMember>
                </soap:Body>
            </soap:Envelope>'''
            
            headers = {
                'Content-Type': 'text/xml; charset=utf-8',
                'SOAPAction': 'UpdateMember'
            }
            
            response = requests.post(
                f"{portal_url}/services/v1_1/VidyoPortalAdminService",
                data=soap_envelope,
                headers=headers,
                auth=HTTPBasicAuth(admin_user, admin_pass),
                timeout=30
            )
            
            return response.status_code == 200 and 'OK' in response.text
            """
            
            # Placeholder simulation
            import time
            time.sleep(1)  # Simulate API call delay
            return True
            
        except Exception as e:
            self.log_message(f"API Error: {str(e)}")
            return False

def main():
    root = tk.Tk()
    app = VidyoPasswordManager(root)
    root.mainloop()

if __name__ == "__main__":
    main()